### ISSUE TITLE
Unhelpful warning for `act` for react-dom@16.8

### ISSUE THREAD

**Issue by ncphillips** [facebook/react#14769]
**Do you want to request a *feature* or report a *bug*?**

Feature/Improvement

**What is the current behavior?**

If there is test code that should be wrapped in `act(...)` then the current warning is given:

```
 console.error node_modules/react-dom/cjs/react-dom.development.js:506
    Warning: An update to null inside a test was not wrapped in act(...).

    When testing, code that causes React state updates should be wrapped into act(...):

    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */

    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://fb.me/react-wrap-tests-with-act
```

When upgrading a large code base, this is basically useless.

**What is the expected behavior?**

Provide at least the test name if not the line number of code that triggered the warning.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

react@16.8.0
react-dom@16.8.0

**Comment by bvaughn** [comment]
> Provide at least the test name if not the line number of code that triggered the warning.

`console.error` should include the line numbers.

**Comment by bvaughn** [comment]
> When upgrading a large code base

Follow-up question: Does this imply that you have a large codebase making use of hooks?

**Comment by ncphillips** [comment]
It is happening, but the error is in react-dom

![image](https://user-images.githubusercontent.com/824015/52356392-802d4200-2a0a-11e9-87e5-df2fbed11355.png)

> Follow-up question: Does this imply that you have a large codebase making use of hooks?

Yup. At least I think the codebase is large, maybe just medium 🤷‍♂️. There's about 800 `/.tsx?/` files. We (perhaps foolishly) have been using the alpha version in production for the last month and have been experimenting with hooks in a variety of places. It just so happens that one of those places is a widely used component that is being rendered in a lot of tests.

**Comment by bvaughn** [comment]
I wouldn't expect that. Do other `console.error` logs have their call stacks truncated as well?

**Comment by threepointone** [comment]
Could you also share your testing setup? Do you use enzyme/react-testing-library? How do you do async testing?

**Comment by gaearon** [comment]
"An update to null" looks weird. Can you try to create a reduced reproducing test case?

**Comment by gaearon** [comment]
Also, what test runner are you using? Can you create a minimal repro?

**Comment by ncphillips** [comment]
We're using `jest` with `react-testing-library`. 

I'll try to carve out some time later to get repro.

**Comment by threepointone** [comment]
react testing library just released an update that integrates act by default with its render/fireEvent calls, so it should pass a bunch of your tests. Any remaining tests probably have explicit mutations, and you should wrap them with act() calls

**Comment by ncphillips** [comment]
I did forget to update `react-testing-library` however upgrading to `5.5.1` did not fix the warnings. Maybe it did fix some, but not all. I don't have a count of how many warnings there were from running the tests. 

Note: all our tests _do_ pass, they just get those warnings.

**Comment by threepointone** [comment]
Without act calls, it's likely your tests might not even be accurate. I updated a whole bunch of fb code for the fb codebase and it wasn't too bad - went through every warning, went to the test that the stack pointed to, and wrapped relevant calls with act() (and caught a couple of bugs in the process)

I'm curious why your stack is getting clipped.

**Comment by FredyC** [comment]
I think this issue is more about making that warning more useful. I've been fighting with same thing yesterday. For a reference have a look [at travis output](https://travis-ci.org/mobxjs/mobx-react-lite/builds/489613981). It's also Jest + react-testing-library.

There is simply no means on figuring out what piece of code is causing that warning because Jest shows only a line where the console.error was produced.

**Comment by ghengeveld** [comment]
For the record, here's how I was testing the `useAsync` hook (simplified):

```js
const Async = ({ children, ...props }) => {
  const renderProps = useAsync(props)
  return children(renderProps)
}

test("returns render props", async () => {
  const promiseFn = () => Promise.resolve("done")
  const component = <Async promiseFn={promiseFn}>{({ data }) => data || null}</Async>
  const { getByText } = render(component)
  await waitForElement(() => getByText("done"))
})
```

`useAsync` calls a [whole bunch of different hooks](https://github.com/ghengeveld/react-async/blob/master/src/useAsync.js) underneith.

**Comment by bvaughn** [comment]
> Please provide decent documentation on how to write tests for custom hooks.

I know it's a bit annoying at the moment, but please be patient with us. We're working on this (both the docs, and the underlying issue).

**Comment by ghengeveld** [comment]
Good to know, and sorry for the harsh-sounding comment, you're doing an excellent job.

**Comment by gaearon** [comment]
If anything, being demanding in issues is a good way to _discourage_ work on something. For example I've spent a lot of effort on this documentation (about a month of work), and comments like this make me not want to even open it anymore. Just something to keep in mind.

**Comment by gaearon** [comment]
@ncphillips We'd still appreciate a reproducing case to see how you got `null` in the warning message.

**Comment by ncphillips** [comment]
Sorry @gaearon it's proving difficult to reproduce 🤔 

It's hard to identify which tests are causing the warning, because they're all passing and since I don't have a warning count, I wouldn't even know if I fixed one. 

I'm starting up a repo now to try and reproduce, but I don't have much time today.

**Comment by panjiesw** [comment]
I created a simple repro for the warning message here https://github.com/panjiesw/act-warning. It uses `jest` (from CRA) and `react-testing-library` v5.5.3

In that repro, the test passed but the warning is still there and only showed component's name, without stack trace and line number

The component code:

```jsx
import React, { useReducer, useEffect } from 'react';

function reducer(state, action) {
  switch (action.type) {
    case 'value':
      return {
        ...state,
        value: action.payload,
      };
    case 'isBar':
      return {
        ...state,
        isBar: action.payload,
      };
    default:
      return state;
  }
}

const Input = () => {
  const [state, dispatch] = useReducer(reducer, { value: '', isBar: 'no' });

  useEffect(() => {
    let mounted = true;
    Promise.resolve().then(() => {
      if (mounted) {
        dispatch({
          type: 'isBar',
          payload: state.value === 'bar' ? 'yes' : 'no',
        });
      }
    });
    return () => {
      mounted = false;
    };
  }, [state.value]);

  const onChange = e => {
    dispatch({ type: 'value', payload: e.target.value });
  };

  return (
    <div>
      <input data-testid="foo" value={state.value} onChange={onChange} />
      <div data-testid="check">isBar: {state.isBar}</div>
    </div>
  );
};

export default Input;
```

The test:

```jsx
import React from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import Input from './Input';

afterEach(cleanup);

jest.useFakeTimers();

test('Input using hooks', done => {
  const w = render(<Input />);

  fireEvent.change(w.getByTestId('foo'), {
    target: { value: 'bar' },
  });

  process.nextTick(() => {
    expect(w.getByTestId('foo').value).toEqual('bar');
    expect(w.getByTestId('check').textContent).toEqual('isBar: yes');
    done();
  });
});

```

**Comment by hisapy** [comment]
Hi devs,

In my case, I'm testing a simple component that `useState` with `mount` from enzyme. 

Notice that I'm not testing the hooks. I'm just `mount`ing the component and simulating some `onChange` and `onClick` and then asserting what props changed. 

Everything passes except that I get the warning:

```
console.error ../../node_modules/react-dom/cjs/react-dom.development.js:506
    Warning: An update to _default inside a test was not wrapped in act(...).

    When testing, code that causes React state updates should be wrapped into act(...):

    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */

    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://fb.me/react-wrap-tests-with-act
```

In the docs it says:

>To reduce the boilerplate, we recommend using react-testing-library which is designed to encourage writing tests that use your components as the end users do.

What does it means for enzyme users? I have a ton of tests that are working but a ton of these warnings as well ... Perhaps, is there a way to silence them?

**Comment by ghengeveld** [comment]
It's not just Enzyme, you'd have the same issue with react-testing-library right now. The team is aware of the issue and working on improving the situation, but having just launched a revolutionary new API means they are probably scrambling to fix a whole lot of other issues as well. It's frustrating if you want to keep up to date with the latest and greatest, but that's how it is. For now I'm just going to disable the affected tests and fly with it.

**Comment by swapkats** [comment]
How could I disable this warning. I use jest and enzyme. I understand the implications but would really like to suppress this warning as it causes unnecessary scrolling and sort of throws me off my workflow.

**Comment by hisapy** [comment]
I added this to my test files

```javascript
// Silence until enzyme fixed to use ReactTestUtils.act()
// If not silenced we get the following warning:
// console.error../../ node_modules / react - dom / cjs / react - dom.development.js: 506
// Warning: An update to _default inside a test was not wrapped in act(...).

// When testing, code that causes React state updates should be wrapped into act(...):

// act(() => {
//   /* fire events that update state */
// });
// /* assert on the output */

// This ensures that you're testing the behavior the user would see in the browser. Learn more at https://fb.me/react-wrap-tests-with-act

const originalError = console.error;

beforeAll(() => {
  console.error = jest.fn();
});

afterAll(() => {
  console.error = originalError;
});

describe("onSubmit login", () => {
  // tests here
})
```

If you want to suppress the warnings _globally_ instead of file by file, you might want to try the above in a jest setup file

**Comment by gaearon** [comment]
Sorry folks — again, we’ll be looking into different scenarios soon and adjust the heuristics.

If you have a test suite full of warnings immediately after a new API that has previously been called an *alpha* was released then I’m sure you can tolerate living with warnings in tests for a few more days. :-)

>but would really like to suppress this warning

We **do not** recommend to silence it hoping that Enzyme will fix them for you.

In some cases, you will need to actually wrap *your* code into `act()` sections. Again, we’ll need to adjust the heuristics and provide more guidance. But I would avoid writing large test suites around APIs while they are in alpha versions if you can’t tolerate some warnings or aren’t sure how to suppress them.

**Comment by flucivja** [comment]
Hello, the problem I have is that even I wrap code in act I receive the warning and I am not sure how to wrap code to avoid it (I tried several combinations without luck).

Here is gist I created to reproduce issue: https://gist.github.com/flucivja/36a324f183c212d0e1cf449631e725a3

At the first I though it is enzyme issue but then I did same without it and warning was still there.

**Comment by gaearon** [comment]
> Do we need to use `act()` only on components using hooks?

It is best practice to always use it — but with classes it's too late to warn due to existing code. That ship has sailed. With Hooks, we have a chance to enforce best practice right from the beginning. Of course, testing code written during alphas would need to adopt, but that's the cost of using alphas.

* Should we consider the code running after `act()` as sync?

Yes. Nothing waits for execution, `act` just defers React mutations until the nested callback exits, and then makes them happen in a batch.

**Comment by marcinczenko** [comment]
I know that people in this thread understand what is the problem, but playing (or being) stupid myself, I want to make sure that we have some super basic example, independent from all project-specific snippets people paste here and there. If I am not on the wrong foot, this should be a representative example of what causes the __act__ warning in the corresponding test"

**Main.js**
```javascript
import React, { useState, useEffect } from 'react'

const Main = () => {
  const [value, setValue] = useState(0)

  useEffect(() => {
    const timeout = setTimeout(() => {
      setValue(1) // <=== this is causing the warning, and it should not
    }, 1000)
    return () => {
      clearTimeout(timeout)
    }
  })

  return (
    <div>
      <p>Current value is: {value}</p>
    </div>
  )
}

export { Main }
```

**Main.test.js**
```javascript
import React from 'react'
import { render, waitForElement } from 'react-testing-library'
import { Main } from './Main'

describe('Main', () => {
  it('is quite annoying', async () => {
    const { getByText } = render(<Main />)

    await waitForElement(() => getByText(/value is: 1/i))
  })
})
```

react-testing-library says the problem is on react side (https://github.com/kentcdodds/react-testing-library/issues/285#issuecomment-461833739) and they refer this thread.

Again, I am just trying to make sure I am looking at the right place, and maybe as a side effect it will help people that just discovered this problem and are traveling between enzyme, react-testing-library, and react, to find the cure.

**Comment by threepointone** [comment]
I was literally writing out these details in a separate post, but I'm typing this out here to relieve some of y'all's anxiety :) 

So first, the immediate, really really short term 'workarounds' if you can't hold on for a couple of days - 
@marcinczenko (and to everyone else, really) in your case, you should be able to pass tests by using fake timers. This is how fb does all its tests and should work in your case. 
```jsx
import React from 'react'
import { render, waitForElement } from 'react-testing-library'
import { Main } from './Main'
jest.useFakeTimers()
describe('Main', () => {
  it('is not so annoying', () => {
    const { getByText } = render(<Main />)
    act(() => { jest.runAllTimers() })
    let element = getByText(/value is: 1/i)
  })
})
```

What of promises then? Even when using promises, if you can mock the returned promise's as something like 
```js
{
  then : jest.fn((callback) => act(() => { callback(mockData) }
}
```

Then, triggering an update would 'set state' in the same 'tick' as you triggered it, and your tests would pass.

However, when using `async`/`await`, javascript resolves its promises in a task queue separate of control from user land. (Happy to explain this in more detail separately, but take my word for it right now) So examples like kent's - https://github.com/kentcdodds/react-testing-library/issues/281#issuecomment-461150694 would be hard to test. The 'fix' for this is heavy
- polyfill promise with https://www.npmjs.com/package/promise, 
- transpile your javascript with `@babel/plugin-transform-runtime` with `regenerator: true`
- then you can use `jest.runAllTimers()` like above to resolve promises and async/await chains too. 

Again, this is how we do it in fb, and it works across all our tests.

Good so far. But we can do a lot better. The above suggestions aren't a satisfying solution for many, so I'm working on a fix right now. We will probably allow async logic inside an `act(...)` block. THIS EXAMPLE IS ONLY A SKETCH, DETAILS MAY CHANGE - 
```jsx
act(() => {
  render(<Main />)
  await waitForElement(() => getByText(/value is: 1/i))
}
```
That should make it a much better experience. The batching behaviour is slightly less enforced, but it should do for most usecases, and we'll have a chance to make it much better with concurrent mode. 

Fair? Happy? I'm trying to get this in fairly soon, so it shouldn't be painful for too long; days, not weeks. Promise. 

(There's a separate issue where warnings aren't showing stack traces for a lot of people, and I'll fix that separately)

**Comment by marcinczenko** [comment]
Thanks @threepointone. It is nice and educative. I have no problem to wait for the ultimate solution, I just started to get lost with the flood of messages and who is fixing what ;).
Here and there I will try the workarounds. Unfortunately there are cases where I cannot mock the async call when component mounts (I mean, I can technically, but this would break the purpose of some of my tests - so I do not want). And heavy fix is fine if one cannot wait, but that feels like an overkill (in nextjs I can easily control babel configuration, but in create-react-app not unless I eject or fork react-scripts or use customize-cra, etc). So for those cases I will be happy to wait.

**Comment by threepointone** [comment]
I think we could have done a better job of communicating this, but I got caught up shipping this internally and let this slip through. I apologise for that.

**Comment by j13l** [comment]
Thanks for the suggested solutions @threepointone. 

Without wanting to rush things, but out of curiosity: should `act()` also do its magic with components wrapped in context?

I created a small example where `act()` doesn't seem to do anything, but if I manually call `update()` on a wrapped component, it works.

Would be happy to get feedback: https://codesandbox.io/s/n5yrrxy6y4

**Comment by gaearon** [comment]
As the first improvement, the 16.8.2 release adds a component stack trace to each message so you can see which component is causing it.

We’ll have more to share when the async version is ready (it’s not yet). This might take another week or so to resolve.

**Comment by quasicomputational** [comment]
I also had some difficulty figuring out just what I was meant to do, especially because the link in the message pointed to the docs of a module I wasn't even using. Some concrete suggestions for additions to the docs that I think would've been helpful for me:

* The [`TestRenderer` docs](https://reactjs.org/docs/test-renderer.html) do not mention `act` once, despite it being exported by that module. The only place to find out about `act` is the ['Test Utilities'](https://reactjs.org/docs/test-utils.html) page or the console error.

* Is batching multiple things in a single `act` call safe, or do they need to be broken up?

* What exactly needs to be wrapped in `act`, even if only as a best practice? Does `TestRenderer.update`? `TestRenderer.create`? But then how would this pattern work:

```
    const renderer = TestRenderer.create(<Swizzy />);
    doSomeAssertions(renderer.toJSON());
```

* Lastly, I don't actually have any idea what `act` does, concretely. It's a piece of magic that I have to put in or else I get big red complaints from the computer. The entire description is currently "To prepare a component for assertions, wrap the code rendering it and performing updates inside an act() call. This makes your test run closer to how React works in the browser." If I had a better mental model of just what `act` was doing, maybe some of the above questions would be easily resolvable for me.

**Comment by threepointone** [comment]
@quasicomputational 
- we're missing docs on `TestRenderer.act`, I'll get it in soon, but the behaviour is exactly like `TestUtils.act`. I'm glad the console error pointed you to the right place, but of course we could do better. 
- yes, batching multiple things in an `act` call is safe. in fact it exposes more bugs. 
- "What exactly needs to be wrapped in `act`?" yes, your `.update` call, `.create` call, timers advancing, promises resolving... anything that could 'change' whatever's on screen, probably needs to wrapped in an `.act` call
- `act` guarantees 2 things 
  - that the updates fired during its callback will be executed and committed to the dom, 
  - any enqueued effects will be flushed. 

You can read some of the original motivation at https://github.com/facebook/react/pull/14744

**Comment by quasicomputational** [comment]
Thanks for the elucidation. I hope that these points can make it into the docs, so that future people won't have to puzzle it out by trial and error. I'd definitely recommend mentioning `act` prominently in the documentation for `update` and `create`, since users will be looking at those anyway. There are also some worked examples on the `TestRenderer` page which don't use `act`.

If I can bikeshed the description, I think that `act`'s documentation should say something like "This will flush React's event queue before it returns, ensuring that your tests do not see an inconsistent state.", which I think should let people understand where it would be needed (assuming I've understood things correctly, of course).

Also, doesn't `act` solve #14050? That's quite useful if so, and the good news should be more widely known.

Concering `TestRenderer.create`, I asked specifically because `act` doesn't allow you to return a value from it (`Warning: The callback passed to TestRenderer.act(...) function must not return anything. You returned: [object Object]`). If I want to inspect the rendered tree and make assertions, those should be outside the function passed to `act` (right?), but I must have access to the `TestRenderer` in order to make those assertions. Is there something I am missing here, or will I need to do something like, e.g.,

```
    const renderer = TestRenderer.create(null);
    act(() => renderer.update(<Swizz />));
    assertions(renderer.toJSON());
```
 or perhaps

```
    let renderer;
    act(() => { renderer = TestRenderer.create(<Swizz />) });
    assertions(renderer.toJSON());
```

in just about every test? This is somewhere where I was looking to the example code for guidance and coming up empty.

**Comment by threepointone** [comment]
- the latter is the pattern we prefer internally. 
```jsx
    let renderer;
    act(() => { renderer = TestRenderer.create(<Swizz />) });
    assertions(renderer.toJSON());
```
alternately, you can make a helper function that does it for you. 

- Yes, `act` does solve #14050! I'll drop a comment there. 

The docs will get better very soon, thank you for the feedback!

**Comment by cellog** [comment]
could a temp workaround for the async issue be to mock useState to return a setState that is wrapped in act? Or are there issues I should think about before trying this? To be clear about my question, I'm trying to test a custom hook that calls an async function in `useEffect` and calls the setState function returned from `useState` inside the async function.

**Comment by threepointone** [comment]
@cellog have you tried the steps in https://github.com/facebook/react/issues/14769#issuecomment-462528230? I wouldn't recommend mocking `setState` in that manner.

**Comment by cellog** [comment]
@threepointone thanks for the fast reply. I am hoping to avoid a major configuration change for the tests, so I guess I'll make the code test-aware (pass in a callback I can use to wrap my setState calls with act, basically, and then run assertions in the callback) until act understands async.

**Comment by cellog** [comment]
@threepointone I came across an elegant solution to this problem. If the async functions are defined in a separate file, using factory functions to pass in the variables from the hook for the closure, you can mock them out and then test the async portion completely separately. No mess, no weird promise polyfills. I think hooks should not have any async stuff directly defined in the hook file for this reason.

**Comment by FredyC** [comment]
@threepointone Just want to note, that `act` warning in 16.8.2 is pretty much unhelpful too :) Yes, there is a call stack leading to the component, but considering there is usually one test file per a component... If it would be only possible to somehow locate the line that caused the unguarded update, that would be awesome :)

Besides, I kinda cannot wrap my head around why there are actual warnings when tests are passing 🤔Doesn't that imply false positives in regards that `act` is not needed there?

**Comment by bvaughn** [comment]
> Besides, I kinda cannot wrap my head around why there are actual warnings when tests are passing 🤔Doesn't that imply false positives in regards that `act` is not needed there?

The warnings are a signal that you may not be testing what you think you're testing. This could be important in cases where passing tests could be false positives.

**Comment by FredyC** [comment]
> The warnings are a signal that you may not be testing what you think you're testing. This could be important in cases where passing tests could be false positives.

Whoa. I've revised some tests that had this issue and indeed there was a bug 😮. After fixing it warning is gone even without `act`. I am starting to like this very much ❤️

**Comment by FredyC** [comment]
Well, so I've encountered one case where I unable to figure why is the warning shown. Unfortunately, CodeSandbox behaves differently here and works just fine. If someone is willing to give it a try, you can download the project and run it locally. There is only a single test which is passing, but a warning is shown.

https://codesandbox.io/s/812mvpzm60

**Comment by threepointone** [comment]
@FredyC wrapping `clock.tick` with `act()` is indeed the recommendation!

**Comment by threepointone** [comment]
I wrote a doc that explains how to use `act(...)` in various scenarios https://github.com/threepointone/react-act-examples. It includes examples and the babel setup for harder cases. I'm still editing working on this, but it could be useful to you folks in this group. Please take some time out to read the document and the included test file. Happy to hear feedback if you have any!

**Comment by FredyC** [comment]
> @FredyC wrapping `clock.tick` with `act()` is indeed the recommendation!

Ok, but I would like to know why :) Given that test is passing, what is `act` helping me with in that case? Isn't it false positive as I've mentioned above? Perhaps it can serve to improve heuristics. It wouldn't be a very helpful tool if we have to be wrapping things with it kinda randomly without any clue.

I've read the whole doc, definitely most appreciated, it's very educative ❤️ Even though it doesn't answer my question it seems.

**Comment by threepointone** [comment]
A single `runAllTimers()` inside an `act()` call might seem like a false positive, but consider if you had many updates happening within that time span, and if these updates trigger new effects too. Without the `act()` call, you wouldn't be guaranteed that the UI has updated to show what a user would see. That would be a buggy test! The very first example in the doc demonstrates this.

**Comment by cellog** [comment]
one quick caveat/question:

```js
    it("can handle async/await", () => {
      // this is the only test that requires the fancy babel setup
      global.Promise = require("promise");
      let resolve;
      function fetch() {
        return new Promise(_resolve => {
          resolve = _resolve;
        });
      }
```

Would there be any potential negative considerations if the `global.Promise = require('promise')` line were in a `jest.setup.js`?

**Comment by threepointone** [comment]
> Would there be any potential negative considerations if the `global.Promise = require('promise')` line were in a `jest.setup.js`?

Not at all, in fact that seems like a cleaner method. I only inlined it here to keep it simple.

**Comment by FredyC** [comment]
> A single `runAllTimers()` inside an `act()` call might seem like a false positive, but consider if you had many updates happening within that time span, and if these updates trigger new effects too. Without the `act()` call, you wouldn't be guaranteed that the UI has updated to show what a user would see. That would be a buggy test! The very first example in the doc demonstrates this.

Well, I am either extra dumb today as I don't get it. The situation is before I wrap `clock.tick` to `act`. I got the test passing and yet there is a warning. I am trying to wrap my head around if either test is wrong or it's false positive on `act` side. I would like to understand this so I can write more tests without guessing what's actually correct.

**Comment by cellog** [comment]
@threepointone one suggestion: The `.babelrc` currently is a global setup. Perhaps modifying it so that it only loads the regenerator stuff in test env would be wise, as otherwise people will be replacing their async/await with the massive and slow regenerator polyfill setup. I also still think it's worth suggesting that another solution is that users separate their async stuff from their hooks so that the actual async activities can be separately tested. A (relatively) simple example is here: https://github.com/unlock-protocol/unlock/pull/1557

**Comment by threepointone** [comment]
@FredyC I'll have a closer look at your test, I only glanced and assumed that's what you needed. codesandbox is odd when it comes to timers and such, so I wouldn't trust it. I'll try locally later sometime.

**Comment by threepointone** [comment]
I hope the doc is helpful to you folks. I'm going to keep working on the document and examples to be clearer, I'll ping back here when I feel it's significantly different.

**Comment by pelotom** [comment]
Is it a reasonable option to mock `setInterval` to wrap its callback in an `act`, or am I somehow defeating the purpose of the warning by not calling `act` explicitly from my test?

```ts
const setInterval = global.setInterval;
global.setInterval = jest.fn((callback: () => void, delay: number) =>
  setInterval(() => act(callback), delay),
);
```

**Comment by marcinczenko** [comment]
Nice doc. I cannot really recall anymore why my experience with `regenerator` is overly bad, but many of my problems with babel bundling and deploying ended when I finally and consistently started to get rid of it from all my babel configurations...

```
[
    '@babel/preset-env',
    {
      exclude: ['transform-regenerator'],
      targets: '> 0.25%, not dead'
    }
]
```

So, I feel really, really uneasy thinking I would have to add it back just to make the tests pass... Yes, I suffer from regenerator trauma. Thus, still waiting for async version of act.

**Comment by cellog** [comment]
Sorry, I think I followed the directions. The simplest test fails with async/await.

[![Edit x9wopqr6xz](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/x9wopqr6xz)

If you have a chance, can you take a look and tell me where I went wrong? Thanks

**Comment by gaearon** [comment]
I’m not sure if you read the above thread but async/await is *not* handled yet. Otherwise the issue would be closed. :-) Stay tuned.

**Comment by cellog** [comment]
Yes I know but there is a hack to enable it using the promise polyfill and regenerator which also didn't work for me. I am staying tuned though.

I wonder if the basic issue is that we should ultimately be using resources and cache for the data fetching and not hooks? But that's for another thread perhaps

**Comment by gaearon** [comment]
In the longer term, yes, but t

**Comment by zanerock** [comment]
I feel like 90% confident that the answer is "yes", but I gather from the thread that the 'async/wait' case will be fixed at some point, so it would be reasonable to pass on the polyfill + babel transform fix from @threepointone excellent doc for the moment and wait for an update?

**Comment by gaearon** [comment]
I don’t suggest anyone to use a polyfill + transform. It doesn’t seem like a fair solution to a testing problem.

**Comment by pelotom** [comment]
Hey, I really appreciate the ongoing efforts to document `act`! If I can make a suggestion, a clear example of a real-world bug that `act` catches would be awesome.

About the warning, my lingering impression is that it's more trouble than it's worth right now, and as such I wish it was opt-in (possibly part of `StrictMode`?), or at least that it was possible to opt out. Having a warning which can't be disabled and which is pointing out a problem for which there is in general no recommended fix* is not great IMO.

\* until we have `act(async () => {...})`, which is "nowhere near being ready"

**Comment by jgoz** [comment]
I have a couple `act()` discoveries, but first I wanted to thank everyone who is working on the Hooks implementation & documentation. I ignored the recommendation and rewrote a bunch of helper components to use hooks 😅... the increase in clarity and composability is even better than what I hoped for after reading the initial post.

**1** &mdash; this may have been obvious to others, but it slipped past me. If you are using Redux and are invoking `store.dispatch` _in your tests_ (and those tests are rendering components that use hooks), those calls must be wrapped in `act`.

We do this enough in our codebase that it necessitated a simple function that wraps a store's `dispatch` method in `act` directly on the store instance. Of course, this would only be used for stores created inside of a test.

**2** &mdash; we also make significant use of `rxjs` in our codebase. Hooks have been incredible here, since subscription and cleanup can be made reusable without HOCs, wrapper components, or manual lifecycle wireup.

Many of our tests use virtual time via a `TestScheduler`. Any events raised inside of a scheduled action (i.e., via `scheduler.schedule(() => clickButton(), 0)`) will trigger the `act` warning. For some reason, it is not enough to wrap the action body in `act` &mdash; I had to subclass `TestScheduler`, overload its `schedule` function, and use `act` there. Something like this:

```js
class ReactTestScheduler extends TestScheduler {
    schedule(work, delay, state) {
        return super.schedule(
            function(s) {
                act(() => { work.call(this, s); });
            },
            delay,
            state
        );
    }
}
```

Using that scheduler implementation, our tests pass without warnings and I even uncovered some incorrect assertions exposed by the increase in strictness. 💪

**Comment by gaearon** [comment]
>* until we have `act(async () => {...})`, which is "nowhere near being ready"

I'm afraid you misunderstood me. "Nowhere near being ready" was referring to *Suspense cache* in response to https://github.com/facebook/react/issues/14769#issuecomment-464741437.

**The `act(async () => {...})` thing is already implemented** in a PR but needs to go through the review process. We are currently in the process of fixing first production bugs in the release. I hope you can agree fixing them and getting those fixes out is a higher priority. As soon as it's done (within a few days), we'll look at the async act warning.

--

At this point we have enough information about the requirements around relaxing this warning. This thread is creating a lot of pressure at the time where we believe addressing more urgent bugs that actually affect production users is more valuable. Since this thread also attracts misunderstandings and misinformation, I'm going to temporarily lock it.

Once we get out 16.8.3 with more urgent fixes, we will look at all different scenarios already described in this thread, and will come up with an API or a relaxed warning that addresses them. I think we already gathered enough feedback here that extra comments won't be necessary in the next few days. Once we have a solution, we will unlock this thread, present it here, and hopefully release that change.

**TLDR: Wait a week or two for this to be fixed. If you managed to write production code with an unstable API while it was in an alpha release, you should also have no troubles temporarily disabling a warning in a test suite.**

Thanks.

**Comment by gaearon** [comment]
Just a small update on this — we’ve needed to refactor some parts of our own test suite to make those changes. And also needed to spend some time on how this will tie into the future Concurrent Mode. But we’re close to having a solution and expect to merge a PR soon. From there we’ll release a canary. The change will go into the next minor.

**Comment by gaearon** [comment]
Another follow up. There’s still ongoing work on this every day. The solution ended up being more wide-reaching than we initially thought, and so we’re making sure it covers all cases before we finalize it. Thanks for bearing with these warnings! I think you’ll like the upcoming API.

**Comment by gaearon** [comment]
https://github.com/facebook/react/pull/14853 has been merged into master and should solve the remaining use cases. We’ll document it together with the next minor release.

**Comment by threepointone** [comment]
https://github.com/facebook/react/releases/tag/v16.9.0-alpha.0 

I'll add docs soon, but I hope it's obvious enough. tldr - 
```jsx
await act(async() => {
 // ... your code that 'does' stuff
})
// your assertions 
```

**Comment by rohitshetty** [comment]
Thanks a lot for your hard work guys! I am facing this error message running my tests. I am new to React ecosystem, can anyone point me to resources which I can use to update my react version for a setup created using CRA? My `"react-scripts": "2.1.8"` and is unejected.

**Comment by gaearon** [comment]
@rohitshetty please share a minimal reproducing project and we can look

**Comment by threepointone** [comment]
https://github.com/facebook/react/issues/14769#issuecomment-461896777 - 
@panjiesw - I took your example, fixed it with the new api, and passed tests 
https://github.com/panjiesw/act-warning/pull/1


https://github.com/facebook/react/issues/14769#issuecomment-462356802 - 
@flucivja if you'd given me a codesandbox/git repo, but if I'm reading it right, this should pass your tests 

```jsx
it('warns that I should use wrap code in act', async () => {
  const deferred = new Deferred(); // https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Promise.jsm/Deferred#Backwards_and_forwards_compatible_helper
  const wrapper = mount(<TestComponent promise={deferred.promise} />); // using enzyme mount
  expect(wrapper.text()).toEqual('{"isLoading":true,"data":null}');
  // async act!
  await act(() => {
    deferred.resolve('test')
    await deferred.promise;
  })        
  expect(wrapper.text()).toEqual('{"isLoading":false,"data":"test"}');
});
```


https://github.com/facebook/react/issues/14769#issuecomment-462518871 - 
@marcinczenko - this snippet should pass your tests. 
```jsx
await act(async () => {
  await waitForElement(() => getByText(/value is: 1/i))
})
```

but also, in the near future, rtl will wrap waitForElement with act, so you won't even need to do the above and it should Just Work™️


https://github.com/facebook/react/issues/14769#issuecomment-464559125 
@cellog your codesandbox link was a bit broken, but I managed to fix the dependencies and such to make the tests pass https://codesandbox.io/s/m4xjz9k06p tl;dr, replacing -
```jsx
rtl.act(() => {
  setNetwork(5);
  jest.runOnlyPendingTimers();
});
```

with this 
```jsx
await act(async() => {
  setNetwork(5);
});
```

passed your tests 


--- 

I'll write a longer explainer next week that'll go into detail about why and how of act(), but I hope you're all unblocked now. And please re-enable console warnings if you'd disabled them :)

**Comment by cellog** [comment]
wow, above and beyond, thanks so much! Sorry for the broken link

**Comment by panjiesw** [comment]
Thank you so much @threepointone 

Will check it out later!

**Comment by FredyC** [comment]
Just a little side note, instead of doing this...
```js
await act(async () => {
  await waitForElement(() => getByText(/value is: 1/i))
})
```
...we can just do this just to go easy on async/await stuff.
```js
await act(() => waitForElement(() => getByText(/value is: 1/i)))
```
Might be even worth to make your own wrapper to increase readability.
```js
await waitForElementWithAct(() => getByText(/value is: 1/i))
```
It's a matter of taste of course ... 🙊

**Comment by threepointone** [comment]
`act()` doesn't let you return a result from the callback, so `await act(() => waitForElement(() => getByText(/value is: 1/i)))` would give another warning. react-testing-library will integrate act() into itslef, so you won't need to make a separate helper. ie, in the very near future, 
```jsx
await waitForElement(() => getByText(/value is: 1/i))
```
will just work, without any warnings.

**Comment by flytaly** [comment]
But in the code `await act(async () => {}) ` async callback is just a normal function that returns promise. How can `act` distinguish async function and callback (like `waitForElement`) that also return promise? It's literally the same thing in JavaScript.

**Comment by threepointone** [comment]
you know what, that's a good point @flytaly . it should actually work lmao :D

**Comment by threepointone** [comment]
BUT you shouldn't expect act() to actually *return* the value from waitForElement. so I'd still not do it just to be explicit. but hey, I'm not your dad.

**Comment by FredyC** [comment]
Is it actually ok to execute expectation within an `act`? I mean if you actually need that return in some cases, do you need to do this?

```
```

Yea, so if you actually need a return value (waitForElement is bad example for that), you have to do this.

**Comment by worc** [comment]
> * use 'act' from 'react-test-renderer' instead of from 'react-dom/test-utils'

this is really key. there's still a reference to the act in test-utils in the docs and it's misleading:

https://reactjs.org/docs/test-utils.html#act

that act doesn't seem to do anything when wrapping a component with hooks.

**Comment by threepointone** [comment]
@worc we'll make the docs better, but it's not exactly hidden there right now 
![image](https://user-images.githubusercontent.com/18808/55973527-ac20aa00-5c7d-11e9-9bb7-f60b0d98598e.png)

**Comment by worc** [comment]
@threepointone but it *doesn't* act the same way. it *works* while the test-utils one doesn't.

**Comment by msteitle** [comment]
I am on react-test-library 6.1.2, wrapping a function component w/ hooks in `act` from react-test-library, and I still get the error from react-dom. Are we just waiting for react-dom 16.9 to resolve these?

*The simple function component:*
```javascript
function UsersPanelProvider(props) {
    const [activePath, setActivePath] = useState();
    const fetchAndSetActivePath = async () => {
        setActivePath(await getActivePath()) // getActivePath returns  Promise.resolve([...])
    }

    useEffect(() => {
        fetchAndSetActivePath();
    }, []);

    return activePath ? <Breadcrumb items={activePath} /> : null;
}
```

*The simple test:*
```javascript
it('should render the panel with contents', async () => {
    let rendered;

    act(() => {
        rendered = render(< UsersPanelProvider {...defaultProps} />);
    });

    const { container, getByText } = rendered;

    expect(container).toMatchSnapshot();
});
```

**Comment by gaearon** [comment]
@worc

>but it doesn't act the same way. it works while the test-utils one doesn't.
 
I don’t know what you mean by this.

Act from `react-test-renderer` works, but only with `react-rest-renderer`. 

Act from `react-dom/test-utils` also works, but only with `react-dom`. 

They both work. But you need to use the one that matches your renderer package.

**Comment by gaearon** [comment]
@msteitle In your example, setting state is asynchronous. Yes, as mentioned in this thread a few times, you’ll need to wait for 16.9.0 or use a canary (instructions above).

**Comment by msteitle** [comment]
Yes, I should have specified, my test utilizes `import { render, act } from 'react-testing-library'`. Thank you for confirming that. I'll hold tight for the react-dom update.

**Comment by zhangkaiyulw** [comment]
Hello, 
Seems updating to 16.9.0-alpha.0 doesn't fix this error in `'react-test-renderer'`.

Now there are double errors besides that is
```
It looks like you wrote TestRenderer.act(async () => ...) or returned a Promise from it's callback. Putting asynchronous logic inside TestRenderer.act(...) is not supported.
```
The code is
```js
    const TestRenderer = require('react-test-renderer');
    await TestRenderer.act(async () => {
      history.back();
      await nextTick();
      history.forward();
      await nextTick();
    });
```

**Comment by zhangkaiyulw** [comment]
Just encountered this issue and after updating to 16.9.0-alpha.0, the error disappears.

Thanks. 👍

**Comment by UsamaHameed** [comment]
Is this still an issue? I have react@16.9.0-alpha.0 installed but `act` only accepts `() => void | undefined`
@threepointone @gaearon

**Comment by FredyC** [comment]
@UsamaHameed You should report that at https://github.com/DefinitelyTyped/DefinitelyTyped repo. The React team does not maintain the Typescript types.

**Comment by andrecalvo** [comment]
Hey all, 

I'm pretty new to hooks so it's possible I'm doing something very wrong. I've tried some of the suggestions on this thread but can't seem to get tests to work. 
I get warnings like: 
`Warning: Do not await the result of calling TestRenderer.act(...), it is not a Promise.`
`When testing, code that causes React state updates should be wrapped into act(...):`

Here's the custom hook and test example (using the alpha react version as suggested):

```
import { useEffect, useState, useRef } from "react";

export const useDataApi = () => {
  const firstUpdate = useRef(true);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(false);
  const [apiUrl, setApiUrl] = useState("");
  const [data, setData] = useState(null);

  useEffect(() => {
    if (!firstUpdate.current) {
      fetch(apiUrl)
        .then(res => res.json())
        .then(response => {
          setIsLoading(false);
          setData(response.reviews);
        })
        .catch(error => {
          setError(error);
        });
    }
    firstUpdate.current = false;
  }, [apiUrl]);

  return [{ data, isLoading, error }, setApiUrl];
};
```

```
import React from "react";
import { useDataApi } from "../../../app/javascript/src/hooks/useDataApi.jsx";
import "whatwg-fetch";
import { renderHook } from "@testing-library/react-hooks";
import fetchMock from "fetch-mock";
import { act } from "react-test-renderer";

describe("useDataApi", () => {
  beforeAll(() => {
    global.fetch = fetch;
  });
  it("should call the api bla bla", async () => {
    const { result } = renderHook(() => useDataApi());
    const [{ data, isLoading, error }, setApiUrl] = result.current;

    await act(() => {
      fetchMock.mock("test.com", {
        reviews: [
          {
            category: "testing"
          },
          {
            category: "testing"
          }
        ]
      });
      setApiUrl("test.com");
    });
  });
});
```

**Comment by threepointone** [comment]
@andrecalvo Could you try replacing `await act(() => {` with `await act(async () => {`

**Comment by andrecalvo** [comment]
> @andrecalvo Could you try replacing `await act(() => {` with `await act(async () => {`

@threepointone unfortunately the same error.
Does this give any more clues? `    Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.`

**Comment by ljharb** [comment]
i believe act in 16.8 is both unable to return a value and unable to be awaited; i don’t think you have much of an option.

**Comment by threepointone** [comment]
Could you share a codesandbox repro? Hard to say whether you’re using the alpha correctly

**Comment by andrecalvo** [comment]
@threepointone what is the correct/different way to use the alpha?

**Comment by 526117** [comment]
ในวันที่ อ. 25 มิ.ย. 2019 02:38 น. Andre <notifications@github.com>
เขียนว่า:

**Comment by 526117** [comment]
ในวันที่ จ. 24 มิ.ย. 2019 23:31 น. Jordan Harband <notifications@github.com>
เขียนว่า:

**Comment by 526117** [comment]
ในวันที่ จ. 24 มิ.ย. 2019 22:01 น. Andre <notifications@github.com>
เขียนว่า:

**Comment by cif** [comment]
@FredyC You can also put this somewhere in your compiler path as a temp fix until DefinitelyTyped is updated:

```js
import { DebugPromiseLike } from 'react-test-renderer';
export function act(callback: () => Promise<void> | undefined): DebugPromiseLike | {};
```

**Comment by andrecalvo** [comment]
> Could you share a codesandbox repro? Hard to say whether you’re using the alpha correctly

Hey @threepointone, here's a Codesandbox. If you check the console you'll see the error I get when I run this on my project. You'll notice I've commented out the fetchMock stuff as it didn't seem to get along well within the Codesandbox environment, it's irrelevant anyway as you can still see the error I'm getting. 

https://codesandbox.io/embed/react-async-fetch-hook-test-failing-6wpkz

**Comment by FredyC** [comment]
@andrecalvo That's easy, you forgot to use `react-test-renderer@16.9.0-alpha.0` ;)

**Comment by andrecalvo** [comment]
@FredyC Aha, I was using the wrong version of react-test-renderer. Updating that but also using 
`waitForNextUpdate();` after my act fixed my issue.

**Comment by meza** [comment]
I'm getting the same error with react-test-renderer when the hook has an async function inside that it awaits on.

**Comment by andyrichardson** [comment]
So I've tried out the latest alpha and the added functionality is great! 

It still doesn't look to cover use cases where async actions are being triggered inside a component on update. I'm assuming the only option is to call a setTimeout when the promise being called can't be accessed inside of a test?

**Comment by threepointone** [comment]
Do you have an example repro? Either as a git repo or codesandbox

**Comment by andyrichardson** [comment]
@threepointone I don't have an example using the latest alpha but here's the current workaround we're using.

 - [setTimeout in test to wait for async action](https://github.com/FormidableLabs/urql/blob/master/src/hooks/useQuery.test.tsx#L138)
 - [async action triggered inside an effect](https://github.com/FormidableLabs/urql/blob/master/src/hooks/useQuery.ts#L76)

**Comment by LeeCheneler** [comment]
Temporary workaround for people who can't upgrade to an alpha version of React like me:

```
/**
 * Suppress React 16.8 act() warnings globally.
 * The react teams fix won't be out of alpha until 16.9.0.
 */
const consoleError = console.error;
beforeAll(() => {
  jest.spyOn(console, 'error').mockImplementation((...args) => {
    if (!args[0].includes('Warning: An update to %s inside a test was not wrapped in act')) {
      consoleError(...args);
    }
  });
});
```

**Comment by mcmillion** [comment]
This did seem to be working (not spitting out warnings) in the alpha until recently? Had to install deps again on a new laptop and now I'm getting warnings again in 16.9.0.alpha.0

**Comment by gaearon** [comment]
We just published `0.0.0-db3ae32b8` of all packages. It is likely going to be a release candidate for 16.9.0 so you can give it a try if you'd like before it gets promoted to stable.

**Comment by pauljfx** [comment]
Just updated my app to react `16.9.0-rc.0` and `16.9.0-rc.0` but still seeing this warning.

This is my code:

```
// useAsyncHook.ts

import { useState } from 'react';

export const someApi = {
  get: (): Promise<string> => {
    return new Promise(resolve => {
      setTimeout(() => {
        resolve(`Hello`);
      }, 1000);
    });
  }
};

export const useAsyncHook = () => {
  const [ data, setData ] = useState<string>('');

  const fetch = async () => {
    const data = await someApi.get();
    setData(data);
  };

  return { data, fetch };
};
```
```
// useAsyncHook.test.ts

import { act, renderHook } from '@testing-library/react-hooks';
import { useAsyncHook } from './useAsyncHook';

describe('My Hook', () => {
  it('Works', async () => {

    const { result, waitForNextUpdate } = renderHook(() => useAsyncHook());

    expect(result.current.data).toEqual('');

    act(() => {
      result.current.fetch();
    });

    await waitForNextUpdate();

    expect(result.current.data).toEqual('Hello');

  });
});
```
The test passes but the warning is still comes up:
```
PASS  src/common/hooks/useApi/useAsyncHook.test.ts
  My Hook
    ✓ Works (1002ms)

  console.error node_modules/react-test-renderer/cjs/react-test-renderer.development.js:102
    Warning: An update to TestHook inside a test was not wrapped in act(...).
    
    When testing, code that causes React state updates should be wrapped into act(...):
    
    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */
    
    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://fb.me/react-wrap-tests-with-act
        in TestHook
        in Suspense

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        1.314s, estimated 2s
Ran all test suites matching /src\/common\/hooks\/useApi\/useAsyncHook.test.ts/i.
```

**Comment by threepointone** [comment]
16.9 got released, including async act, and updated documentation https://reactjs.org/blog/2019/08/08/react-v16.9.0.html Closing this, cheers.

**Comment by mharidy** [comment]
upgrading react and react-dom to 16.9 doesn't fix the problem, the only change console.error node_modules/react-dom/cjs/react-dom.development.js:**545**  insted of 
 console.error node_modules/react-dom/cjs/react-dom.development.js:**506**

**Comment by FredyC** [comment]
@mharidy See #16348 where I tried to improve that and now it continues over at https://github.com/facebook/jest/issues/8819

**Comment by ernestohegi** [comment]
https://reactjs.org/blog/2019/08/08/react-v16.9.0.html#async-act-for-testing this solved all my problems when using custom **async** hooks. No more `act` warnings and tests are passing!

**Comment by pavanshinde47** [comment]
Can you please share an example of your test?

**Comment by gre** [comment]
is there a way to actively disable this log feature with React?

It is not helpful for case like when you run tests from inside a React app. I've wrote a "jest-like" runner inside React Native that run my tests. the test does not involve React testing at all! but the fact the tests are run FROM react creates these crazy logs.

as a "test runner implementer" how do i disable this log to happen.

Thanks.

<img width="476" alt="Capture d’écran 2019-09-13 à 14 30 03" src="https://user-images.githubusercontent.com/211411/64862509-fd6d0500-d632-11e9-965c-ee2600db5c22.png">

**Comment by jktravis** [comment]
I had a couple tests that would do this, and had a really hard time finding the problem. But for me, I was using react-testing-library, I just added an extra `await wait()` at the bottom of each test that was giving me the problem.  

Maybe you could implement an inherit wait for every test?

**Comment by threepointone** [comment]
There's no way to disable the warning in your scenario, you could try patching console.error. 

> I had a couple tests that would do this, and had a really hard time finding the problem. But for me, I was using react-testing-library, I just added an extra await wait() at the bottom of each test that was giving me the problem.
>
> Maybe you could implement an inherit wait for every test?

I would *not* recommend this. @jktravis I'd recommend instead fixing the test and wrapping the section generating the async update with `await act(async () => ...)`

**Comment by mpopv** [comment]
@jktravis This is similar to how I solved this issue as well -- I think it (in light of @threepointone's reply above) reveals a sort of tension between how async `act` is intended to be used and how `testing-library` implements it internally.

`act` is meant to be wrapped around operations that _themselves_ trigger a rerender -- you wait for an async function to resolve when all state changes finish, then test the results. So when I upgraded, I expected tests to look like this:

```
// example of what NOT to do
const { getByText } = await render(<Todos/>);
expect(getByText('Todos: 1')).toBeTruthy();
await fireEvent.click(getByText('Add Todo'));
expect(getByText('Todos: 2')).toBeTruthy();
```

But `testing-library` sort of does the opposite -- you fire methods that trigger state changes, then wait for an async function that resolves when some query returns true. It's the _query_ that's awaited, rather than the operation that triggers the rerender. So it's `wait` that internally wraps `act`, rather than `fireEvent`, etc.:

```
const { getByText } = render(<Todos/>);
await wait(() => getByText('Todos: 1'));
expect(getByText('Todos: 1')).toBeTruthy();
fireEvent.click(getByText('Add Todo'));
await wait(() => getByText('Todos: 2'));
expect(getByText('Todos: 2')).toBeTruthy(); 
```

I only stumbled upon this while trying to figure out why I was still seeing `act` warnings after rewriting it the first way.

**Comment by pelotom** [comment]
How can `fireEvent.click` return a promise if it doesn't know when all the changes it kicked off have actually completed? Clicking a button could trigger a fetch which asynchronously updates your component some time later; there's no way for `fireEvent.click` to know that.

**Comment by jedmao** [comment]
I'm still getting this error with react-scripts@3.3.0 (recently published) and React v16.12.0. Specifically, I'm getting it with React Testing Library, whenever I call a `useReducer`-generated `dispatch` within a `useEffect`.

See [this RTL issue](https://github.com/testing-library/react-testing-library/issues/535#issuecomment-563819446) for more context.

**Comment by rahulcyadav** [comment]
@threepointone could you help fix this test
https://codesandbox.io/s/jest-enzyme-re41k

**Comment by FredyC** [comment]
@rahulcyadav You should probably ask over at Jest repo. Error is popping from Jest code. Besides, there is no problem with "act" warning. This isn't general "fix my tests" issue :)

**Comment by gaearon** [comment]
Jest 25.4.0 includes a JS stack for `console.error` so this should improve things. They're asking for feedback! https://github.com/facebook/jest/issues/8819#issuecomment-616231071

**Comment by Siddharth77** [comment]
My sincere THANKS to @ernestohegi for guiding with his solution and I did bit refactoring of the same and the thing which has resolved error for me is as below : 
`/* eslint-disable react/prop-types */
import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import { createMount } from '@material-ui/core/test-utils';
import xComponent from './xComponent';

let mount = createMount();

beforeEach(() => {
  mount = createMount({ strict: true });
});

afterEach(() => {
  mount.cleanUp();
});

it('renders without crashing', async () => {
  const div = document.createElement('div');
  await act(async () => {
    ReactDOM.render(<xComponent />, div);
  });
  ReactDOM.unmountComponentAtNode(div);
});`

**Comment by AncientSwordRage** [comment]
I'm still seeing this despite having react 16.13.1 and jest 24.9.0, any idea on how to diagnose?
